storage/store.go

```diff
@@ -1412,7 +1412,45 @@ func (s *Store) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.Bat
 		// logical ticks added here.
 		ba.Timestamp.Forward(s.Clock().Now().Add(0, int32(len(ba.Requests))-1))
 	}
-	s.ctx.Clock.Update(ba.Timestamp)
+	// We update the clock and hold on to the timestamp - we know that any
+	// write with a higher timestamp we run into later must have started
+	// after this point in (absolute) time.
+	now := s.ctx.Clock.Update(ba.Timestamp)
+
+	if ba.Txn != nil {
+		// Reduce uncertainty restarts by attaching the above timestamp to the
+		// returned response or error (if in a txn). The caller can use it to
+		// shorten its uncertainty interval when it comes back to this node.
+		defer func() {
```

Comment:
I think this will get a lot cleaner if you always return `now` in the response header. Then the caller can update its clock and possibly the observed timestamp corresponding to this node if the batch was part of a transaction.